### ISSUE TITLE
Image2styleGAN复现

### ISSUE THREAD

**Issue by youngAt19** [07hyx06/07hyx06.github.io#1]
你好，我最近也在看Image2styleGAN这篇论文，试图复现它，但是因为自己入门不久，一直没有思路。恰好看到了你这篇推文，请问能看一下你的代码学习一下吗？

[https://github.com/07hyx06/07hyx06.github.io/blob/master/cv_blog/0015_Image2StyleGAN.html](url)

**Comment by 07hyx06** [comment]
from config import *
from model.VGG import vgg16
from model.StyleGAN import style_generator
import torch
import torch.nn as nn
from PIL import Image
from torchvision import transforms
import torch.optim as optim
import numpy as np
import torch.nn.functional as F
from utils import *

def train():
    # 预训练的StyleGAN生成器
    _, g_s = style_generator()
    g_s.to(DEVICE)
    g_s.eval()
    freeze_grad(g_s)
    # 预训练的VGG
    vgg = vgg16()
    vgg.to(DEVICE)
    vgg.eval()
    freeze_grad(vgg)
    # 分别提取与StyleGAN比对的图像和输入VGG的图像
    img, resized_img = prepare_img_for_image2StyleGAN()
    img = img.to(DEVICE)
    resized_img = resized_img.to(DEVICE)
    # 获取latent初始值
    latent = get_latent(AVG_LATENT_PATH)
    latent = latent.to(DEVICE)
    latent.requires_grad = True
    # latent = torch.randn(1, 18, 512, device=DEVICE, requires_grad=True)
    # init_latent.requires_grad = True

    criterion = nn.MSELoss().to(DEVICE)
    conv1_1, conv1_2, conv3_2, conv4_2 = vgg(resized_img)
    optimizer = optim.Adam([{'params': latent}], lr=LEARNING_RATE)

    for iter in range(1, 1 + TOTAL_ITER):
        optimizer.zero_grad()
        pred = g_s(latent)
        resized_pred = F.avg_pool2d(pred, 4, 4)
        c11, c12, c32, c42 = vgg(resized_pred)
        l_img = criterion(img, pred)
        l_c11 = criterion(conv1_1, c11)
        l_c12 = criterion(conv1_2, c12)
        l_c32 = criterion(conv3_2, c32)
        l_c42 = criterion(conv4_2, c42)
        loss = l_img + l_c11 + l_c12 + l_c32 + l_c42
        loss.backward()
        optimizer.step()
        if iter % PRINT_FREQ == 0:
            print('iter:' + str(iter))
            print('tot loss:' + str(loss.item()) + '||img loss:' + str(l_img.item()))
            print('c11 loss:' + str(l_c11.item()) + '||c12 loss:' + str(l_c12.item()))
            print('c32 loss:' + str(l_c32.item()) + '||c42 loss:' + str(l_c42.item()))
            print('---------')
    # print(latent.size())
    np.save(REC_LATENT_PATH, latent[0].cpu().detach().numpy())

def image2StyleGAN():
    train()
    present_from_latent(get_latent(REC_LATENT_PATH), REC_IMG_PATH)

if __name__ == '__main__':
    train()
    present_from_latent(get_latent(REC_LATENT_PATH), REC_IMG_PATH)

把这段代码放到https://github.com/07hyx06/FaceLab/tree/master/StyleGAN/net下，再改改import应该就能用了。顺便想问下您是从哪搜到我的博客的:D

**Comment by 07hyx06** [comment]
```
from config import *
from model.VGG import vgg16
from model.StyleGAN import style_generator
import torch
import torch.nn as nn
from PIL import Image
from torchvision import transforms
import torch.optim as optim
import numpy as np
import torch.nn.functional as F
from utils import *


def train():
    # 预训练的StyleGAN生成器
    _, g_s = style_generator()
    g_s.to(DEVICE)
    g_s.eval()
    freeze_grad(g_s)
    # 预训练的VGG
    vgg = vgg16()
    vgg.to(DEVICE)
    vgg.eval()
    freeze_grad(vgg)
    # 分别提取与StyleGAN比对的图像和输入VGG的图像
    img, resized_img = prepare_img_for_image2StyleGAN()
    img = img.to(DEVICE)
    resized_img = resized_img.to(DEVICE)
    # 获取latent初始值
    latent = get_latent(AVG_LATENT_PATH)
    latent = latent.to(DEVICE)
    latent.requires_grad = True
    # latent = torch.randn(1, 18, 512, device=DEVICE, requires_grad=True)

    # init_latent.requires_grad = True

    criterion = nn.MSELoss().to(DEVICE)
    conv1_1, conv1_2, conv3_2, conv4_2 = vgg(resized_img)
    optimizer = optim.Adam([{'params': latent}], lr=LEARNING_RATE)

    for iter in range(1, 1 + TOTAL_ITER):
        optimizer.zero_grad()
        pred = g_s(latent)
        resized_pred = F.avg_pool2d(pred, 4, 4)
        c11, c12, c32, c42 = vgg(resized_pred)
        l_img = criterion(img, pred)
        l_c11 = criterion(conv1_1, c11)
        l_c12 = criterion(conv1_2, c12)
        l_c32 = criterion(conv3_2, c32)
        l_c42 = criterion(conv4_2, c42)
        loss = l_img + l_c11 + l_c12 + l_c32 + l_c42
        loss.backward()
        optimizer.step()
        if iter % PRINT_FREQ == 0:
            print('iter:' + str(iter))
            print('tot loss:' + str(loss.item()) + '||img loss:' + str(l_img.item()))
            print('c11 loss:' + str(l_c11.item()) + '||c12 loss:' + str(l_c12.item()))
            print('c32 loss:' + str(l_c32.item()) + '||c42 loss:' + str(l_c42.item()))
            print('---------')
    # print(latent.size())
    np.save(REC_LATENT_PATH, latent[0].cpu().detach().numpy())


def image2StyleGAN():
    train()
    present_from_latent(get_latent(REC_LATENT_PATH), REC_IMG_PATH)


if __name__ == '__main__':
    train()
    present_from_latent(get_latent(REC_LATENT_PATH), REC_IMG_PATH)
```
这段代码放到[https://github.com/07hyx06/FaceLab/tree/master/StyleGAN/net](url)下就能用了，当然还需要改一下import。顺便想问下您是从哪搜到我的博客的:D

**Comment by youngAt19** [comment]
好，太谢谢你了！！我是在github搜image2styleGAN然后找到的。
![image](https://user-images.githubusercontent.com/58137631/84849003-e9a17400-b086-11ea-81ae-b39d9c53e838.png)

**Comment by youngAt19** [comment]
![image](https://user-images.githubusercontent.com/58137631/85254278-7175ec80-b492-11ea-9307-c09334d0f746.png)
请问在这个prepare的过程中你是怎么处理的呢？我自己尝试写了一下，但是跑出来的效果并不好
